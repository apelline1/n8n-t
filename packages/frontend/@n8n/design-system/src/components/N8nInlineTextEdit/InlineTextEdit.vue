<script lang="ts" setup>
import {
	EditableArea,
	EditableInput,
	EditablePreview,
	EditableRoot,
	useForwardPropsEmits,
} from 'reka-ui';
import { useTemplateRef } from 'vue';

import type { EditableRootEmits, EditableRootProps } from '../../reka-ui';

const props = withDefaults(defineProps<EditableRootProps>(), {
	maxLength: 100,
	placeholder: 'Enter text...',
	selectOnFocus: true,
	autoResize: false,
	submitMode: 'both',
});

const emit = defineEmits<EditableRootEmits>();
const forwarded = useForwardPropsEmits(props, emit);

const editableRoot = useTemplateRef('editableRoot');

function forceFocus() {
	if (!props.readonly) {
		editableRoot.value?.edit();
	}
}

function forceCancel() {
	editableRoot.value?.cancel();
}

defineExpose({ forceFocus, forceCancel });

function splitStringByEndLength(str: string, endLength: number): [string, string] {
	if (str.length <= endLength) {
		return [str, ''];
	}
	return [str.slice(0, str.length - endLength), str.slice(-endLength)];
}
</script>

<template>
	<EditableRoot v-bind="forwarded" ref="editableRoot" class="inline-text-edit">
		<EditableArea class="editable-area">
			<EditablePreview class="inline-rename-preview" data-test-id="inline-edit-preview" as="div">
				<template v-if="props.modelValue">
					<div
						v-for="(segment, index) in splitStringByEndLength(props.modelValue, 4)"
						:key="index"
						class="text-clamp"
					>
						{{ segment }}
					</div>
				</template>
			</EditablePreview>
			<EditableInput class="inline-rename-input" data-test-id="inline-edit-input" />
		</EditableArea>
	</EditableRoot>
</template>

<style scoped>
.editable-area {
	border-width: 1px;
	border-style: solid;
	border-color: transparent;
	border-radius: var(--border-radius-base);
	border-color: var(--color-foreground-base);
	padding: 4px;
	max-width: 350px;
	min-width: 64px;
	&:hover:not([data-focused]) {
		border-color: var(--color-foreground-base);
		cursor: pointer;
	}

	&[data-focused] {
		border-color: var(--color-secondary);
	}

	&[data-readonly] {
		pointer-events: none;
		border-color: transparent;
	}
}

.inline-rename-input {
	field-sizing: content;
	max-width: 100%;
	width: 100%;
	border: 0;
	outline: 0;
}

.inline-rename-preview {
	white-space: pre;
	user-select: none;
	padding: 2px 0;
	display: grid;
	grid-template-columns: 1fr auto;
	max-width: 100%;
	&[hidden] {
		display: none;
	}
}

.text-clamp {
	white-space: nowrap;
	min-width: 0;
	overflow: hidden;
	text-overflow: ellipsis;
}
</style>
