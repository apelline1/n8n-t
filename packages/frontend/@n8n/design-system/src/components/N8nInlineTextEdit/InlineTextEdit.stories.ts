import type { StoryFn } from '@storybook/vue3';
import { ref } from 'vue';

import InlineTextEdit from './InlineTextEdit.vue';

export default {
	title: 'Atoms/InlineTextEdit',
	component: InlineTextEdit,
};

const Template: StoryFn = (args, { argTypes }) => ({
	setup() {
		const model = ref(args.modelValue);
		return { args, model };
	},
	props: Object.keys(argTypes),
	components: {
		InlineTextEdit,
	},
	template: `
		<div class="story">
			<N8nInlineTextEdit v-bind="args" v-model="model"/>
		</div>
	`,
});

export const primary = Template.bind({});
primary.args = {
	modelValue: 'Test',
};

export const placeholder = Template.bind({});
placeholder.args = {
	modelValue: '',
	placeholder: 'Enter workflow name',
};
