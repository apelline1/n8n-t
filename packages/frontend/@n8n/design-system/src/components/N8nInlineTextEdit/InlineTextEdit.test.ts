import userEvent from '@testing-library/user-event';

import { createComponentRenderer } from '@n8n/design-system/__tests__/render';

import N8nInlineTextEdit from './InlineTextEdit.vue';

const renderComponent = createComponentRenderer(N8nInlineTextEdit);

describe('InlineTextEditCopy', () => {
	it('should render correctly', () => {
		const wrapper = renderComponent({
			props: {
				modelValue: 'Test Value',
			},
		});

		expect(wrapper.html()).toMatchSnapshot();
	});

	it('should emit value on enter', async () => {
		const wrapper = renderComponent({
			props: {
				modelValue: 'Test Value',
			},
		});
		const input = wrapper.getByTestId('inline-edit-input');

		await wrapper.rerender({ modelValue: 'New Value' });
		await userEvent.clear(input);
		await userEvent.type(input, 'Updated Value');
		await userEvent.keyboard('{Enter}');

		expect(wrapper.emitted('update:modelValue')[0]).toStrictEqual(['Updated Value']);
	});

	it('should not emit value on escape key press', async () => {
		const wrapper = renderComponent({
			props: {
				modelValue: 'Test Value',
			},
		});
		const input = wrapper.getByTestId('inline-edit-input');

		await userEvent.type(input, 'Updated Value');
		await userEvent.keyboard('{Escape}');

		expect(wrapper.emitted('update:modelValue')).toBeFalsy();
	});
});
