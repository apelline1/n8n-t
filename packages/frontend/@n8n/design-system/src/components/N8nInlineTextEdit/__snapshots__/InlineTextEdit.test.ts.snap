// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InlineTextEditCopy > should render correctly 1`] = `
"<div data-v-e54e5004="" class="inline-text-edit" dir="ltr" data-dismissable-layer="">
  <div data-v-e54e5004="" data-placeholder-shown="" style="display: inline-grid;" class="editable-area"><span data-v-e54e5004="" tabindex="0" data-placeholder-shown="" style="white-space: pre; user-select: none; grid-area: 1 / 1 / auto / auto; overflow: hidden; text-overflow: ellipsis;" class="inline-rename-preview" data-test-id="inline-edit-preview">Test Value</span><input data-v-e54e5004="" placeholder="Enter text..." maxlength="100" aria-label="editable input" style="all: unset; grid-area: 1 / 1 / auto / auto; visibility: hidden;" class="inline-rename-input" data-test-id="inline-edit-input" value="Test Value"></div>
  <!---->
</div>"
`;
